File: src/main/java/com/mbayou/kick4k/authorization/AuthorizationClient.java
package com.mbayou.kick4k.authorization;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mbayou.kick4k.KickConfiguration;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.time.Instant;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * OAuth 2.0 Authorization client supporting PKCE flow for the Kick API.
 * <p>
 * Handles generation of authorization URLs, exchanging authorization codes for tokens,
 * refreshing access tokens, and managing token state.
 */
public class AuthorizationClient {
    private final HttpClient httpClient;
    private final ObjectMapper mapper;
    private final KickConfiguration configuration;
    private final RefreshTokenStore refreshToken;

    private String accessToken;
    private Instant expiresAt;

    /**
     * Creates a new AuthorizationClient.
     *
     * @param httpClient   the HttpClient used for HTTP requests
     * @param mapper       ObjectMapper used to parse JSON responses
     */
    public AuthorizationClient(HttpClient httpClient, ObjectMapper mapper, KickConfiguration configuration, RefreshTokenStore refreshToken) {
        this.httpClient = httpClient;
        this.mapper = mapper;
        this.configuration = configuration;
        this.refreshToken = refreshToken;
    }

    /**
     * Generates a cryptographically random code verifier for PKCE.
     * <p>
     * The verifier is a URL-safe base64 encoded string without padding, 32 bytes in length.
     *
     * @return a new code verifier string
     */
    public static String generateCodeVerifier() {
        byte[] code = new byte[32];
        new SecureRandom().nextBytes(code);
        return base64UrlEncoder(code);
    }

    /**
     * Generates a code challenge from the given code verifier using SHA-256 and Base64URL encoding,
     * as specified in the PKCE RFC.
     *
     * @param codeVerifier the code verifier string
     * @return the corresponding code challenge string
     */
    public static String generateCodeChallenge(String codeVerifier) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(codeVerifier.getBytes(StandardCharsets.US_ASCII));
            return base64UrlEncoder(hash);
        } catch (NoSuchAlgorithmException exception) {
            throw new RuntimeException("Failed to generate code challenge", exception);
        }
    }

    /**
     * Builds the OAuth authorization URL where the user should be redirected to grant access.
     *
     * @param scopeList     list of OAuth scopes requested by the application
     * @param codeChallenge the PKCE code challenge
     * @return the full URL string for the authorization request
     */
    public String getAuthorizationUrl(List<Scope> scopeList, String codeChallenge) {
        StringJoiner joiner = new StringJoiner("&",
                this.configuration.getOAuthHost() + this.configuration.getAuthorizationEndpoint() + "?",
                "");
        joiner.add("client_id=" + encodeUrl(this.configuration.getClientId()));
        joiner.add("response_type=code");
        joiner.add("redirect_uri=" + encodeUrl(this.configuration.getRedirectUri()));
        // Kick API requires state parameter but does not use it afterwards – generated here for best practice
        joiner.add("state=" + encodeUrl(UUID.randomUUID().toString()));
        String scopes = scopeList.stream()
                .map(Scope::getScope)
                .collect(Collectors.joining(" "));
        joiner.add("scope=" + encodeUrl(scopes));
        joiner.add("code_challenge=" + encodeUrl(codeChallenge));
        joiner.add("code_challenge_method=S256");

        return joiner.toString();
    }

    /**
     * Exchanges the authorization code received from the OAuth callback and the original
     * code verifier for access and refresh tokens.
     *
     * @param code         the authorization code returned by the OAuth server
     * @param codeVerifier the original PKCE code verifier
     * @return an OAuthTokenResponse containing access and refresh tokens
     */
    public OAuthTokenResponse exchangeCodeForToken(String code, String codeVerifier) {
        Map<String, String> body = Map.of(
                "code", code,
                "client_id", this.configuration.getClientId(),
                "client_secret", this.configuration.getClientSecret(),
                "redirect_uri", this.configuration.getRedirectUri(),
                "grant_type", "authorization_code",
                "code_verifier", codeVerifier
        );

        return this.postOAuthTokenRequest(body);
    }

    /**
     * Refreshes the access token using the stored refresh token.
     *
     * @return a new OAuthTokenResponse containing refreshed access and refresh tokens
     */
    public OAuthTokenResponse refreshAccessToken() {
        Map<String, String> body = Map.of(
                "refresh_token", this.refreshToken.getRefreshToken(),
                "client_id", this.configuration.getClientId(),
                "client_secret", this.configuration.getClientSecret(),
                "grant_type", "refresh_token"
        );

        OAuthTokenResponse newToken = this.postOAuthTokenRequest(body);
        this.refreshToken.notifyRefreshTokenRoll(newToken.getRefreshToken());
        return newToken;
    }

    /**
     * Returns a valid access token. Automatically refreshes the token if it is missing or expired.
     *
     * @return a valid OAuth access token string
     */
    public String getAccessToken() {
        if (this.accessToken == null || Instant.now().isAfter(this.expiresAt.minusSeconds(10))) {
            OAuthTokenResponse response = this.refreshAccessToken();
            this.setTokens(response);
        }

        return this.accessToken;
    }

    /**
     * Updates the stored tokens and expiry time.
     *
     * @param oAuthToken the OAuthTokenResponse containing the new tokens and expiry
     */
    public void setTokens(OAuthTokenResponse oAuthToken) {
        this.accessToken = oAuthToken.getAccessToken();
        this.refreshToken.notifyRefreshTokenRoll(oAuthToken.getRefreshToken());
        this.expiresAt = Instant.now().plusSeconds(oAuthToken.getExpiresIn());
    }

    /**
     * Sets the access token manually.
     *
     * @param accessToken new access token
     */
    public void setAccessToken(String accessToken) {
        this.accessToken = accessToken;
    }

    /**
     * Sets the expiration time of the current access token.
     *
     * @param expiresAt expiration timestamp
     */
    public void setExpiresAt(Instant expiresAt) {
        this.expiresAt = expiresAt;
    }

    private OAuthTokenResponse postOAuthTokenRequest(Map<String, String> body) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(this.configuration.getOAuthHost() + this.configuration.getTokenEndpoint()))
                    .header("Content-Type", "application/x-www-form-urlencoded")
                    .POST(HttpRequest.BodyPublishers.ofString(generateUrlencodedForm(body)))
                    .build();

            HttpResponse<String> response = this.httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new OAuthTokenException(response.statusCode(), response.body());
            }

            return this.mapper.readValue(response.body(), OAuthTokenResponse.class);
        } catch (IOException | InterruptedException exception) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Failed to create OAuth token", exception);
        }
    }

    private static String generateUrlencodedForm(Map<String, String> data) {
        StringJoiner formData = new StringJoiner("&");

        for (Map.Entry<String, String> entry : data.entrySet()) {
            formData.add(encodeUrl(entry.getKey()) + "=" + encodeUrl(entry.getValue()));
        }

        return formData.toString();
    }

    private static String encodeUrl(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static String base64UrlEncoder(byte[] input) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(input);
    }
}

File: src/main/kotlin/com/mbayou/kick4k/authorization/AuthorizationClient.kt
package com.mbayou.kick4k.authorization

import com.fasterxml.jackson.databind.ObjectMapper
import com.mbayou.kick4k.KickConfiguration
import java.io.IOException
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.time.Instant
import java.util.Base64
import java.util.StringJoiner
import java.util.UUID

class AuthorizationClient(
    private val httpClient: HttpClient,
    private val mapper: ObjectMapper,
    private val configuration: KickConfiguration,
    private val refreshToken: RefreshTokenStore,
) {
    private var accessTokenValue: String? = null
    private var expiresAt: Instant = Instant.EPOCH

    fun getAuthorizationUrl(scopeList: List<Scope>, codeChallenge: String): String {
        val joiner = StringJoiner("&", "${configuration.oAuthHost}${configuration.authorizationEndpoint}?", "")
        joiner.add("client_id=${encodeUrl(configuration.clientId)}")
        joiner.add("response_type=code")
        joiner.add("redirect_uri=${encodeUrl(configuration.redirectUri)}")
        joiner.add("state=${encodeUrl(UUID.randomUUID().toString())}")
        val scopes = scopeList.joinToString(" ") { it.scope }
        joiner.add("scope=${encodeUrl(scopes)}")
        joiner.add("code_challenge=${encodeUrl(codeChallenge)}")
        joiner.add("code_challenge_method=S256")
        return joiner.toString()
    }

    fun exchangeCodeForToken(code: String, codeVerifier: String): OAuthTokenResponse {
        val body = mapOf(
            "code" to code,
            "client_id" to configuration.clientId,
            "client_secret" to configuration.clientSecret,
            "redirect_uri" to configuration.redirectUri,
            "grant_type" to "authorization_code",
            "code_verifier" to codeVerifier,
        )
        return postOAuthTokenRequest(body)
    }

    fun refreshAccessToken(): OAuthTokenResponse {
        val body = mapOf(
            "refresh_token" to refreshToken.getRefreshToken(),
            "client_id" to configuration.clientId,
            "client_secret" to configuration.clientSecret,
            "grant_type" to "refresh_token",
        )

        val newToken = postOAuthTokenRequest(body)
        refreshToken.notifyRefreshTokenRoll(newToken.refreshToken)
        return newToken
    }

    fun getAccessToken(): String {
        val current = accessTokenValue
        val now = Instant.now()
        if (current == null || now.isAfter(expiresAt.minusSeconds(10))) {
            val response = refreshAccessToken()
            setTokens(response)
            return response.accessToken
        }
        return current
    }

    fun setTokens(oAuthToken: OAuthTokenResponse) {
        accessTokenValue = oAuthToken.accessToken
        refreshToken.notifyRefreshTokenRoll(oAuthToken.refreshToken)
        expiresAt = Instant.now().plusSeconds(oAuthToken.expiresIn.toLong())
    }

    fun setAccessToken(accessToken: String) {
        this.accessTokenValue = accessToken
    }

    fun setExpiresAt(expiresAt: Instant) {
        this.expiresAt = expiresAt
    }

    private fun postOAuthTokenRequest(body: Map<String, String>): OAuthTokenResponse {
        val request = HttpRequest.newBuilder()
            .uri(URI.create(configuration.oAuthHost + configuration.tokenEndpoint))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .POST(HttpRequest.BodyPublishers.ofString(generateUrlencodedForm(body)))
            .build()

        val response = try {
            httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        } catch (exception: InterruptedException) {
            Thread.currentThread().interrupt()
            throw RuntimeException("Failed to create OAuth token", exception)
        } catch (exception: IOException) {
            throw RuntimeException("Failed to create OAuth token", exception)
        }

        if (response.statusCode() != 200) {
            throw OAuthTokenException(response.statusCode(), response.body())
        }

        return mapper.readValue(response.body(), OAuthTokenResponse::class.java)
    }

    companion object {
        fun generateCodeVerifier(): String {
            val code = ByteArray(32)
            SecureRandom().nextBytes(code)
            return base64UrlEncoder(code)
        }

        fun generateCodeChallenge(codeVerifier: String): String {
            return try {
                val digest = MessageDigest.getInstance("SHA-256")
                val hash = digest.digest(codeVerifier.toByteArray(StandardCharsets.US_ASCII))
                base64UrlEncoder(hash)
            } catch (exception: NoSuchAlgorithmException) {
                throw RuntimeException("Failed to generate code challenge", exception)
            }
        }
    }
}

private fun generateUrlencodedForm(data: Map<String, String>): String {
    val formData = StringJoiner("&")
    data.forEach { (key, value) -> formData.add("${encodeUrl(key)}=${encodeUrl(value)}") }
    return formData.toString()
}

private fun encodeUrl(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8)

private fun base64UrlEncoder(input: ByteArray): String =
    Base64.getUrlEncoder().withoutPadding().encodeToString(input)
